File: frameworks/custom-applet/sdk-ts/src/protocol.ts
export type IconKind = "name" | "path";

export class Icon {
  constructor(
    public readonly kind: IconKind,
    public readonly value: string,
  ) {}

  static name(value: string): Icon {
    return new Icon("name", value);
  }

  static path(value: string): Icon {
    return new Icon("path", value);
  }

  toProtocol(): { name: string } | { path: string } {
    return this.kind === "name" ? { name: this.value } : { path: this.value };
  }
}

export class StatusItem {
  constructor(
    public readonly options: {
      id?: string;
      icon?: Icon;
      label?: string;
      tooltip?: string;
    } = {},
  ) {}

  toProtocol(): Record<string, unknown> {
    const payload: Record<string, unknown> = {};
    if (this.options.id !== undefined) {
      payload.id = this.options.id;
    }
    if (this.options.icon !== undefined) {
      payload.icon = this.options.icon.toProtocol();
    }
    if (this.options.label !== undefined) {
      payload.label = this.options.label;
    }
    if (this.options.tooltip !== undefined) {
      payload.tooltip = this.options.tooltip;
    }
    return payload;
  }
}

export class MenuItem {
  constructor(
    public readonly options: {
      id: string;
      label: string;
      visible?: boolean;
      enabled?: boolean;
    },
  ) {}

  toProtocol(): Record<string, unknown> {
    const payload: Record<string, unknown> = {
      id: this.options.id,
      label: this.options.label,
    };
    if (this.options.visible !== undefined) {
      payload.visible = this.options.visible;
    }
    if (this.options.enabled !== undefined) {
      payload.enabled = this.options.enabled;
    }
    return payload;
  }
}

File: frameworks/custom-applet/sdk/glimpse_applet/protocol.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(slots=True)
class Icon:
    kind: str
    value: str

    @classmethod
    def name(cls, value: str) -> "Icon":
        return cls(kind="name", value=value)

    @classmethod
    def path(cls, value: str) -> "Icon":
        return cls(kind="path", value=value)

    def to_protocol(self) -> dict[str, str]:
        return {self.kind: self.value}


@dataclass(slots=True)
class MenuItem:
    id: str
    label: str
    visible: bool | None = None
    enabled: bool | None = None

    def to_protocol(self) -> dict[str, object]:
        payload: dict[str, object] = {
            "id": self.id,
            "label": self.label,
        }
        if self.visible is not None:
            payload["visible"] = self.visible
        if self.enabled is not None:
            payload["enabled"] = self.enabled
        return payload


@dataclass(slots=True)
class StatusItem:
    id: str | None = None
    icon: Icon | None = None
    label: str | None = None
    tooltip: str | None = None

    def to_protocol(self) -> dict[str, object]:
        payload: dict[str, object] = {}
        if self.id is not None:
            payload["id"] = self.id
        if self.icon is not None:
            payload["icon"] = self.icon.to_protocol()
        if self.label is not None:
            payload["label"] = self.label
        if self.tooltip is not None:
            payload["tooltip"] = self.tooltip
        return payload
